wave_lang/kernel/wave/mlir_converter/water_emitter.py

```diff
@@ -0,0 +1,282 @@
+        root_graph: fx.Graph = trace.get_root_graph()
+
+        # Collect placeholders from root graph
+        placeholders = [
+            n
+            for n in trace.walk()
+            if getattr(n, "op", "") == "placeholder"
+            and getattr(n, "graph", None) is root_graph
+        ]
+
+        # Build function argument types from placeholders
+        # TODO: This might need special handling when iterate nodes are present
+        arg_types = []
+        for p in placeholders:
+            c = get_custom(p)
+            t = getattr(c, "_type", None) or getattr(c, "type", None)
+            arg_types.append(_type_to_wave_mlir(ctx, t))
+
+        # Return type of the function is always empty
+        func_type = ir.FunctionType.get(arg_types, [])
+        with ir.InsertionPoint(module.body):
+            func_op = func.FuncOp("kernel", func_type)
+            # TODO: WaveHyperparameterAttr only supports int currently, so we
+            #       lose mappings like ADDRESS_SPACE_A: SHARED_ADDRESS_SPACE
+            # Convert the symbols in subs to strings and attach as
+            # WaveHyperparameterAttr to func_op
+            func_op.operation.attributes["wave.hyperparameters"] = (
+                wave.WaveHyperparameterAttr.get(
+                    {str(k): v for k, v in options.subs.items() if isinstance(v, int)}
+                )
+            )
+            entry_block = ir.Block.create_at_start(func_op.regions[0], arg_types)
+
+            # Map placeholders to function arguments
+            for i, fx_node in enumerate(placeholders):
+                value_map[fx_node] = entry_block.arguments[i]
+
+            with ir.InsertionPoint(entry_block):
+                # Emit in original order to preserve dependencies
+                for fx_node in root_graph.nodes:
+                    node = get_custom(fx_node)
+                    # Ensure types are inferred
+                    node.infer_type()
+
+                    # No MLIR ops are emitted for placeholder and output nodes
+                    if isinstance(node, Placeholder | Output):
+                        continue
+
+                    # Collect already emitted mlir values for the args of this node
+                    operands = [
+                        mlir_arg
+                        for arg in fx_node.args
+                        if (mlir_arg := value_map.get(arg)) is not None
+                    ]
+                    result_type = _type_to_wave_mlir(ctx, node.type)
+
+                    mlir_op = None
+                    if node.tkw_op_name in WAVE_OP_CONSTRUCTORS:
+                        op_builder = WAVE_OP_CONSTRUCTORS[node.tkw_op_name]
+                        # TODO: Add special handling for Iterate node
+                        if isinstance(node, Write):
+                            mlir_op = op_builder(
+                                value_map[node.register_], value_map[node.memory]
+                            )
```

Comment:
Why did we need to special-case write? It would be useful to have documentation at the top saying which ops need special-casing.